Detectors/TOF/compression/src/Compressor.cxx

```diff
@@ -0,0 +1,1496 @@
+      o2::raw::HBFUtils::printRDH(*rdh);
+    }
+#endif
+
+    auto headerSize = rdh->headerSize;
+    auto memorySize = rdh->memorySize;
+    auto offsetToNext = rdh->offsetToNext;
+    auto drmPayload = memorySize - headerSize;
+
+    /** copy DRM payload to save buffer **/
+    std::memcpy(mDecoderSaveBuffer + mDecoderSaveBufferDataSize, reinterpret_cast<char*>(rdh) + headerSize, drmPayload);
+    mDecoderSaveBufferDataSize += drmPayload;
+
+    /** move to next RDH **/
+    rdh = reinterpret_cast<o2::header::RAWDataHeader*>(reinterpret_cast<char*>(rdh) + offsetToNext);
+
+    /** check next RDH is within buffer **/
+    if (reinterpret_cast<char*>(rdh) < mDecoderBuffer + mDecoderBufferSize)
+      continue;
+
+    /** otherwise return **/
+    return true;
+  }
+
+#ifdef DECODER_VERBOSE
+  if (mDecoderVerbose) {
+    std::cout << colorBlue
+              << "--- RDH close detected"
+              << colorReset
+              << std::endl;
+    o2::raw::HBFUtils::printRDH(*rdh);
+  }
+#endif
+
+  /** copy RDH open to encoder buffer **/
+  std::memcpy(mEncoderPointer, mDecoderRDH, mDecoderRDH->headerSize);
+  mEncoderPointer = reinterpret_cast<uint32_t*>(reinterpret_cast<char*>(mEncoderPointer) + rdh->headerSize);
+
+  /** process DRM data **/
+  mDecoderPointer = reinterpret_cast<uint32_t*>(mDecoderSaveBuffer);
+  mDecoderPointerMax = reinterpret_cast<uint32_t*>(mDecoderSaveBuffer + mDecoderSaveBufferDataSize);
+  while (mDecoderPointer < mDecoderPointerMax) {
+    mEventCounter++;
+    if (processDRM())
+      break;
+  }
+  mDecoderSaveBufferDataSize = 0;
+
+  /** bring encoder pointer back if fatal error **/
+  if (mDecoderFatal) {
+    mFatalCounter++;
+    mEncoderPointer = mEncoderPointerStart;
+  }
+
+  if (mDecoderError)
+    mErrorCounter++;
+
+  /** updated encoder RDH open **/
+  mEncoderRDH->memorySize = reinterpret_cast<char*>(mEncoderPointer) - reinterpret_cast<char*>(mEncoderRDH);
+  mEncoderRDH->offsetToNext = mEncoderRDH->memorySize;
+
+  /** copy RDH close to encoder buffer **/
+  /** CAREFUL WITH THE PAGE COUNTER **/
+  mEncoderRDH = reinterpret_cast<o2::header::RAWDataHeader*>(mEncoderPointer);
+  std::memcpy(mEncoderRDH, rdh, rdh->headerSize);
+  mEncoderRDH->memorySize = rdh->headerSize;
+  mEncoderRDH->offsetToNext = mEncoderRDH->memorySize;
+  mEncoderPointer = reinterpret_cast<uint32_t*>(reinterpret_cast<char*>(mEncoderPointer) + rdh->headerSize);
+
+#ifdef DECODER_VERBOSE
+  if (mDecoderVerbose) {
+    std::cout << colorBlue
+              << "--- END PROCESS HBF"
+              << colorReset
+              << std::endl;
+  }
+#endif
+
+  /** move to next RDH **/
+  mDecoderPointer = reinterpret_cast<uint32_t*>(reinterpret_cast<char*>(rdh) + rdh->offsetToNext);
+
+  /** check next RDH is within buffer **/
+  if (reinterpret_cast<char*>(mDecoderPointer) < mDecoderBuffer + mDecoderBufferSize)
+    return false;
+
+  /** otherwise return **/
+  return true;
+}
+
+bool Compressor::processDRM()
+{
```

Comment:
Hi Ruben,

this is the core method of the compression, which has to go as fast as possible.
I'm not sure how trivial it is to split it in few more methods, I will have to have a look at it with a bit of patience.

I hope it is fine with you if this can be postponed to a following PR.

Cheers,
Roberto

Reply:
As you prefer, though splitting it to a few functions called w/o any arguments should not lead to any noticeable performance loss.